design-proposals/orch-cli.md

```diff
@@ -0,0 +1,229 @@
+component shall add their own nouns as appropriate. For example,
+
+- `cli list applications` ... The `application` noun applies to applications in the catalog.
+
+- `cli list registry` ... The `registry` noun applies to registries in the catalog.
+
+#### Subjects
+
+Subjects are additional information or context releated to a noun. Different nouns may support
+different subjects. For convenience sake, most nouns support a name subject and optionally for those
+objects that are versioned, a version subject. For example,
+
+- `cli get application nginx` ... returns the first nginx application, agnostic to the version.
+
+- `cli get application nginx 0.0.1` ... return exactly version 0.0.1 of the nginx application.
+
+If additional subjects beyond name and version are required, then they should be specified with addional
+options. For example,
+
+- `cli get application nginx --deployed true` ... get the nginx application that is currently deployed.
+
+#### Global Options
+
+- `-h` / `--help` ... return help. The help returned may be contextualized to the verb or noun that is
+  being used. For example `cli -h` returns global help whereas `cli create application -h` return help
+  that is relevant to creating applications.
+
+- `-n` / `--noauth` ... disables authentication checks, only useful in a devleopment environment.
+
+- `-p` / `--project` ... sets the project that will be used for the current verb. Required for verbs that
+  act on projects.
+
+- `-v` / `--verbose` ... enables verbose output.
+
+- `-o json` / `-o yaml` ... instead of returning human-readable output, return output in `yaml` or `json`.
+
+#### Verb-Specific and/or Noun-Specific options
+
+Verbs and nouns may have addional non-global options as necessary. For example, the `create application`
+verb/noun pair includes the options `--chart-name` and `--chart-registry` and `--chart-version`.
+
+#### Examples
+
+- `cli login sample-project-admin` ... log in as a user.
+
+- `cli -p acme list applications nginx` ... list all nginx applications in a tabular format.
+
+- `cli -p acme list applications nginx =o yaml` ... list all nginx applications and emit the output as yaml.
+
+- `cli -p acme create application nginx 0.0.1 --chart-name nginx --chart-registry bitnami --chart-version 0.0.1`
+  ... create the nginx application using the specified parameters.
+
+- `cli -p acme delete applciation nginx 0.0.1` ... delete the nginx application.
+
+- `cli -p acme update application nginx 0.0.1 --chart-registry dockerhub` ... change the chart-registry for an
+  application.
+
+- `cli -p acme apply -f nginx.yaml` ... create the objects contained inside nginx.yaml.
+
+- `cli logout` ... log out.
+
+### Miscellaneous Guidelines
+
+- The CLI seeks to minimize the boundaries between teams and components. For example, the following approaches
+  are not desirable:
+
+  - `cli update app-orch application foo` ... app-orch is a team/subsystem designator that is not relevant to
```

Comment:
I am wondering if we should consider the potential of adding groups/subgroups like cloud provider CLIs. Let's say we want to perform start/stop VM operations; what does the command look like?"

Reply:
My first thought would be `cli start <vm>` and `cli stop <vm>`. Grabbing the most similar example I can think of, `kubectl cordon node` and `kubectl uncordon node`. We could also fold it into update, such as `cli update <appname> -- running false`